vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go

```diff
@@ -692,17 +701,58 @@ func (l *terminationLoggingListener) Accept() (net.Conn, error) {
 
 	select {
 	case <-l.lateStopCh:
-		klog.Warningf("Accepted new connection from %s very late in the graceful termination process (more than 80%% has passed), possibly a sign for a broken load balancer setup.", c.RemoteAddr().String())
-		if swapped := l.lateConnReceived.CAS(false, true); swapped {
-			l.Eventf(corev1.EventTypeWarning, "LateConnections", "The apiserver received connections (e.g. from %q) very late in the graceful termination process, possibly a sign for a broken load balancer setup.", c.RemoteAddr().String())
-		}
+		lateConnectionRemoteAddrsLock.Lock()
+		defer lateConnectionRemoteAddrsLock.Unlock()
+		lateConnectionRemoteAddrs[c.RemoteAddr().String()] = true
 	default:
-		select {
-		case <-l.stopCh:
-			klog.V(2).Infof("Accepted new connection from %s during graceful termination.", c.RemoteAddr())
-		default:
-		}
 	}
 
-	return c, err
+	return c, nil
+}
+
+// WithLateConnectionFilter logs every non-probe request that comes through a late connection identified by remote address.
+func WithLateConnectionFilter(handler http.Handler) http.Handler {
+	var lateRequestReceived atomic.Bool
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		lateConnectionRemoteAddrsLock.RLock()
+		late := lateConnectionRemoteAddrs[r.RemoteAddr]
+		lateConnectionRemoteAddrsLock.RUnlock()
+
+		if late {
+			ctx := r.Context()
+			requestInfo, ok := apirequest.RequestInfoFrom(ctx)
+			if !ok {
+				err := fmt.Errorf("no RequestInfo found in context, handler chain must be wrong")
+				errorMsg := fmt.Sprintf("Internal Server Error: %#v", r.RequestURI)
```

Comment:
what would you do with an error message like that on the client-side? Could it be more informative?